src/library/features/crates/cratefeature.cpp

```diff
@@ -58,21 +57,21 @@ CrateFeature::CrateFeature(UserSettingsPointer pConfig,
     // if closure does not have the same number of crates as the crates table
     // this means that the user just started mixxx with nested crates for the
     // first time, so we have to fill the closure table with (self,self,0)
-    if (!m_crateHierarchy.closureIsValid()) {
+    if (!pTrackCollection->crates().closureIsValid()) {
         QMessageBox::warning(
             nullptr,
             tr("Nested Crates"),
             tr("Crates now support hierarchical structure."
                "All your crates have been converted to Level 1 crates"));
 
-        m_crateHierarchy.resetClosure();
-        m_crateHierarchy.initClosure();
-        m_crateHierarchy.resetPath();
-        m_crateHierarchy.generateAllPaths();
+        pTrackCollection->crates().resetClosure();
+        pTrackCollection->crates().initClosure();
+        pTrackCollection->crates().resetPath();
+        pTrackCollection->crates().generateAllPaths();
```

Comment:
A single call to TrackCollection should take care of this, including checking closureIsValid. How about calling it TrackCollection::checkCrateHierarchy? I don't think there should be any mention of "closure" in CrateFeature; that's an implementation detail.